src/box/vy_scheduler.c

```diff
@@ -2037,6 +2067,12 @@ vy_scheduler_f(va_list va)
 			 */
 			continue;
 		}
+		if (fiber_is_cancelled()) {
+			if (scheduler->stat.tasks_inprogress == 0)
+				break;
+			fiber_cond_wait(&scheduler->scheduler_cond);
+			continue;
+		}
```

Comment:
I don't like that this fiber doesn't exit on shutdown. Please consider exiting immediately and waiting for tasks completion in `vy_scheduler_shutdown`.

Reply:
Done. The related hunks now looks like below.

<details>
<summary>diff</summary>

```diff
diff --git a/src/box/vy_scheduler.c b/src/box/vy_scheduler.c
index 677bf5749..d413a1708 100644
--- a/src/box/vy_scheduler.c
+++ b/src/box/vy_scheduler.c
@@ -468,15 +484,32 @@ vy_scheduler_start(struct vy_scheduler *scheduler)
        fiber_start(scheduler->scheduler_fiber, scheduler);
 }

+static void
+vy_scheduler_finish_tasks(struct vy_scheduler *scheduler, int *tasks_done,
+                         int *tasks_failed);
+
 void
-vy_scheduler_destroy(struct vy_scheduler *scheduler)
+vy_scheduler_shutdown(struct vy_scheduler *scheduler)
 {
-       /* Stop scheduler fiber. */
+       vy_worker_pool_start_shutdown(&scheduler->dump_pool);
+       vy_worker_pool_start_shutdown(&scheduler->compaction_pool);
+
+       fiber_cancel(scheduler->scheduler_fiber);
+       fiber_join(scheduler->scheduler_fiber);
        scheduler->scheduler_fiber = NULL;
-       /* Sic: fiber_cancel() can't be used here. */
-       fiber_cond_signal(&scheduler->dump_cond);
-       fiber_cond_signal(&scheduler->scheduler_cond);
+       while (true) {
+               int tasks_done, tasks_failed;
+               vy_scheduler_finish_tasks(scheduler, &tasks_done,
+                                         &tasks_failed);
+               if (scheduler->stat.tasks_inprogress == 0)
+                       break;
+               fiber_cond_wait(&scheduler->scheduler_cond);
+       }
+}

+void
+vy_scheduler_destroy(struct vy_scheduler *scheduler)
+{
        vy_worker_pool_destroy(&scheduler->dump_pool);
        vy_worker_pool_destroy(&scheduler->compaction_pool);
        diag_destroy(&scheduler->diag);
@@ -1996,47 +2038,46 @@ vy_task_complete(struct vy_task *task)
        return -1;
 }

+/** Complete and delete all currently processed tasks. */
+static void
+vy_scheduler_finish_tasks(struct vy_scheduler *scheduler, int *tasks_done,
+                         int *tasks_failed)
+{
+       *tasks_done = 0;
+       *tasks_failed = 0;
+       while(!stailq_empty(&scheduler->processed_tasks)) {
+               struct stailq tasks = STAILQ_INITIALIZER(tasks);
+               struct vy_task *task, *next;
+               stailq_concat(&tasks, &scheduler->processed_tasks);
+               stailq_foreach_entry_safe(task, next, &tasks, in_processed) {
+                       if (vy_task_complete(task) == 0)
+                               (*tasks_done)++;
+                       else
+                               (*tasks_failed)++;
+                       vy_worker_pool_put(task->worker);
+                       vy_task_delete(task);
+               }
+       }
+}
+
 static int
 vy_scheduler_f(va_list va)
 {
        struct vy_scheduler *scheduler = va_arg(va, struct vy_scheduler *);

-       while (scheduler->scheduler_fiber != NULL) {
-               struct stailq processed_tasks;
-               struct vy_task *task, *next;
-               int tasks_failed = 0, tasks_done = 0;
+       while (!fiber_is_cancelled()) {
+               struct vy_task *task;
+               int tasks_done, tasks_failed;

                fiber_check_gc();
-               /* Get the list of processed tasks. */
-               stailq_create(&processed_tasks);
-               stailq_concat(&processed_tasks, &scheduler->processed_tasks);
-
-               /* Complete and delete all processed tasks. */
-               stailq_foreach_entry_safe(task, next, &processed_tasks,
-                                         in_processed) {
-                       if (vy_task_complete(task) != 0)
-                               tasks_failed++;
-                       else
-                               tasks_done++;
-                       vy_worker_pool_put(task->worker);
-                       vy_task_delete(task);
-               }
+               vy_scheduler_finish_tasks(scheduler, &tasks_done,
+                                         &tasks_failed);
                /*
                 * Reset the timeout if we managed to successfully
                 * complete at least one task.
                 */
-               if (tasks_done > 0) {
+               if (tasks_done > 0)
                        scheduler->timeout = 0;
-                       /*
-                        * Task completion callback may yield, which
-                        * opens a time window for a worker to submit
-                        * a processed task and wake up the scheduler
-                        * (via scheduler_async). Hence we should go
-                        * and recheck the processed_tasks in order not
-                        * to lose a wakeup event and hang for good.
-                        */
-                       continue;
-               }
                /* Throttle for a while if a task failed. */
                if (tasks_failed > 0)
                        goto error;
```

</details>